providers/implementations/kem/ecx_kem.c

```diff
@@ -0,0 +1,704 @@
+            return 0;
+        mode = ossl_eckem_modename2id(p->data);
+        if (mode == KEM_MODE_UNDEFINED)
+            return 0;
+        ctx->mode = mode;
+    }
+    return 1;
+}
+
+static const OSSL_PARAM known_settable_ecxkem_ctx_params[] = {
+    OSSL_PARAM_utf8_string(OSSL_KEM_PARAM_OPERATION, NULL, 0),
+    OSSL_PARAM_octet_string(OSSL_KEM_PARAM_IKME, NULL, 0),
+    OSSL_PARAM_END
+};
+
+static const OSSL_PARAM *ecxkem_settable_ctx_params(ossl_unused void *vctx,
+                                                   ossl_unused void *provctx)
+{
+    return known_settable_ecxkem_ctx_params;
+}
+
+/*
+ * See Section 4.1 DH-Based KEM (DHKEM) ExtractAndExpand
+ */
+static int dhkem_extract_and_expand(EVP_KDF_CTX *kctx,
+                                    unsigned char *okm, size_t okmlen,
+                                    uint16_t kemid,
+                                    const unsigned char *dhkm, size_t dhkmlen,
+                                    const unsigned char *kemctx,
+                                    size_t kemctxlen)
+{
+    uint8_t suiteid[5];
+    uint8_t prk[EVP_MAX_MD_SIZE];
+    size_t prklen = okmlen; /* Nh */
+    int ret;
+
+    if (prklen > sizeof(prk))
+        return 0;
+
+    ossl_dhkem_getsuiteid(suiteid, kemid);
+
+    ret = ossl_hpke_labeled_extract(kctx, prk, prklen,
+                                    NULL, 0, suiteid, sizeof(suiteid),
+                                    "eae_prk", dhkm, dhkmlen)
+          && ossl_hpke_labeled_expand(kctx, okm, okmlen, prk, prklen,
+                                      suiteid, sizeof(suiteid),
+                                      "shared_secret", kemctx, kemctxlen);
+    OPENSSL_cleanse(prk, prklen);
+    return ret;
+}
+
+/*
+ * See Section 7.1.3 DeriveKeyPair.
+ *
+ * This function is used by ecx keygen.
+ * (For this reason it does not use any of the state stored in PROV_ECX_CTX).
+ *
+ * Params:
+ *     ecx An initialized ecx key.
+ *     privout The buffer to store the generated private key into (it is assumed
+ *             this is of length ecx->keylen).
+ *     ikm buffer containing the input key material (seed). This must be non NULL.
+ *     ikmlen size of the ikm buffer in bytes
+ * Returns:
+ *     1 if successful or 0 otherwise.
+ */
+int ossl_ecx_dhkem_derive_private(ECX_KEY *ecx, unsigned char *privout,
+                                  const unsigned char *ikm, size_t ikmlen)
+{
+    int ret = 0;
+    EVP_KDF_CTX *kdfctx = NULL;
+    unsigned char prk[EVP_MAX_MD_SIZE];
+    uint16_t kemid;
+    const char *kdfdigestname;
+    uint8_t suiteid[5];
+    size_t prklen;
+
+    /* ikmlen should have a length of at least Nsk */
+    if (ikmlen < ecx->keylen) {
+        ERR_raise_data(ERR_LIB_PROV, PROV_R_INVALID_INPUT_LENGTH,
+                       "ikm length is :%zu, should be at least %zu",
+                       ikmlen, ecx->keylen);
+        goto err;
+    }
+
+    if (ecx->type == ECX_KEY_TYPE_X25519) {
+        kemid = 0x20;
```

Comment:
What is this magic number?

Reply:
Shifted to a #define with a description, plus made this code into a function that is shared by the 2 places that use it.